### DISCUSSION TITLE
Convert type DynamicJaxprTracer to DeviceArray

### DISCUSSION THREAD

**Discussion by XDongiang** [jax-ml/jax#13295] [Q&A]
During the calculation process, I found problems with the calculation of these two types of arrays, so I hope that these two tyoes can be converted.
<class 'jax.interpreters.partial_eval.DynamicJaxprTracer'> to <class 'jaxlib.xla_extension.DeviceArray'> and 
<class 'jax.interpreters.ad.JVPTracer'> to <class 'jaxlib.xla_extension.DeviceArray'>

**Comment by jakevdp** [comment]
*Edit: there is now a FAQ entry on this topic: https://jax.readthedocs.io/en/latest/faq.html#how-can-i-convert-a-jax-tracer-to-a-numpy-array*

---

Hi - JAX uses abstract tracers to determine the behavior of functions when transforming them (with e.g. `jit` and `grad`) You can read more about this in [How To Think In JAX](https://jax.readthedocs.io/en/latest/notebooks/thinking_in_jax.html). Tracers cannot be converted to Device Arrays, because in general they are abstractions that do not have knowledge of the contents of the arrays that will eventually be passed to the transformed function.

That said, I think this is an [XY problem](https://en.wikipedia.org/wiki/XY_problem): your goal is not to convert tracers to arrays (which is impossible by design), but rather to accomplish some particular application. If you can edit your question to give more details about what exactly your application is, we may be able to point you to a solution.

**Reply by XDongiang** [reply]
Thank you for your reply!
I use the POT package to calculate the swd distance, where my source and target are both device arrays but the weight used is a tracer object.
(ot.sliced.sliced_wasserstein_distance)[https://pythonot.github.io/gen_modules/ot.sliced.html]
my code is:
```
self.y = device_put(np.array(np.hstack((cdl.data_phi_p_kk,cdl.data_f_p_kk))),device=self.device)
self.x = device_put(np.array(np.hstack((cdl.mc_phi_p_kk,cdl.mc_f_p_kk))),device=self.device)
wt = np.sum(dplex.dabs(np.einsum("mljk->mjk", lasso_data_phif0_kk_BW_flatte980)+np.einsum("mljk->mjk", lasso_data_phif0_kk_BW_BW)+np.einsum("mljk->mjk", lasso_data_phif2_kk_BW_flatte1270)+np.einsum("mljk->mjk", lasso_data_phif2_kk_BW_BW)),axis=1)
return swd(self.x, self.y, a=wt, n_projections=256)
```

**Reply by jakevdp** [reply]
Thanks for the details. It looks like the POT package is not designed for compatibility with JAX. For performance, it would be best to avoid calling external packages and instead implement the particular operations you want to use on top of JAX primitives (or to find a package that does this for you).

Alternatively, you could use [`jax.pure_callback`](https://jax.readthedocs.io/en/jax-v0.3.23/_autosummary/jax.pure_callback.html) to call-out to the POT package from a JAX computation, but be aware that this may lead to severe performance impacts, particularly when executing code on accelerators like GPU.
